### Práctica 3 - Aprendizaje automático
    Miguel Simó Juan
    43194679N
    
El objetivo de la práctica es agrupar y arreglar los datos para que un algoritmo de aprendizaje pueda leerlos y entenderlos
y así cuando el algoritmo reciba un test podrá predecir su valor.
    
En primer lugar importaremos todo lo necesario y utilizaremos panda para leer el dataset:

In [212]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

In [213]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

El proceso que vamos a llevar a cabo es el siguiente:
> - Eliminar las columnas que no aporten ningún tipo de información
> - Buscar los jugadores que no pueden pasar por el algoritmo debido a falta de datos
> - Corregir datos que pueden ser relevantes pero no están expresados de forma correcta para el algoritmo

Bajo estos puntos, lo primero que haremos será eliminar las dos primeras columnas ya que los identificadores no aportan nada.

In [214]:
df = df.iloc[:, 2:]
df.head()

Name  Age                                           Photo  \
0           L. Messi   31  https://cdn.sofifa.org/players/4/19/158023.png   
1  Cristiano Ronaldo   33   https://cdn.sofifa.org/players/4/19/20801.png   
2          Neymar Jr   26  https://cdn.sofifa.org/players/4/19/190871.png   
3             De Gea   27  https://cdn.sofifa.org/players/4/19/193080.png   
4       K. De Bruyne   27  https://cdn.sofifa.org/players/4/19/192985.png   

  Nationality                                 Flag  Overall  Potential  \
0   Argentina  https://cdn.sofifa.org/flags/52.png       94         94   
1    Portugal  https://cdn.sofifa.org/flags/38.png       94         94   
2      Brazil  https://cdn.sofifa.org/flags/54.png       92         93   
3       Spain  https://cdn.sofifa.org/flags/45.png       91         93   
4     Belgium   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       

Acto seguido sacamos de la tabla los datos que son completamente irrelevantes tales como la foto, la bandera, el logo del club...
Asumimos que, además de las columnas que son imágenes o nombres (las cuales lógicamente no aportan ninguna información), las columnas que contienen datos sobre el contrato no tienen impacto en el algoritmo puesto que conocer las fechas de duración de un contrato no hará que un jugador que tenga altas estadísticas se vea afectado y, por ejemplo, un "rookie" puede ser muy bueno o muy malo independientemente de si acaba de entrar en un equipo o no.

In [215]:
df.pop("Photo")
df.pop("Flag")
df.pop("Loaned From")
df.pop("Club Logo")
df.pop("Jersey Number")
df.pop("Name")
df.pop("Real Face")
df.pop("Contract Valid Until")
df.pop("Joined")
df.head()


Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

   Special Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0     2202           Left                       5.0        4.0          4.0   
1     2228          Right                       5.0        4.0          5.0   
2     2143          Right                       5.0        5.0          5.0   
3     1471          Right                       4.0        3.0          1.0   
4     2281          Right                       4.0        5.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
3  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        19.0         51.0         42.0          57.0         58.0     60.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
3       90.0     43.0       31.0     67.0     43.0      64.0       12.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       

El siguiente paso será hacer un descarte de un bajo porcentaje de los jugadores a los cuales no se les puede aplicar el algoritmo debido a falta de datos. 
Mediante la siguiente codificación buscaremos aquellos jugadores que no tienen club, posición o altura. El por qué de ello es simple, estos jugadores representan un porcentaje muy bajo de la muestra, además los que no tienen altura coinciden en su mayoría con jugadores de los cuales no tenemos muchos datos (tales como peso, posición...).


In [216]:
df = df.dropna(subset = ['Club'])
df = df.dropna(subset = ['Position'])
df = df.dropna(subset = ['Height'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

   Special Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0     2202           Left                       5.0        4.0          4.0   
1     2228          Right                       5.0        4.0          5.0   
2     2143          Right                       5.0        5.0          5.0   
3     1471          Right                       4.0        3.0          1.0   
4     2281          Right                       4.0        5.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
3  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        19.0         51.0         42.0          57.0         58.0     60.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
3       90.0     43.0       31.0     67.0     43.0      64.0       12.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       

A continuación vamos a reordenar ciertos datos para que el algoritmo pueda entenderlos. En este caso sabemos que tener una columna de clubes, nacionalidades... No es muy relevante para el algoritmo y que a este no le sirve leer un string con el nombre de un club por ejemplo. Así que nuestro siguiente paso será convertir cada club, nacionalidad, tipo de cuerpo... En una columna distinta y añadiremos un 1 si el jugador pertenece a alguna de esas características o un 0 en su defecto.
Las columnas seleccionadas para este proceso son: club, nacionalidad, pie preferente, tipo de cuerpo, posición y Work Rate.

Durante el proceso he tenido que investigar ciertos aspectos de algunas de las columnas. Un ejemplo de esto es el work rate. El work rate es un valor que se le da al ataque/defensa del jugador según su estilo de juego. Que un jugador tenga un work rate alto/alto no quiere decir que sea mejor que uno que lo tenga medio/medio, simplemente cambia su posición en el campo según si es atacante o defensor. Bajo este criterio he dividido el work rate según como atacan y como defienden. Seguiremos el mismo criterio que con el resto y añadiremos un 1 si el jugador contiene esta característica o un 0 si no.

In [217]:
clb = df.pop("Club")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

nat = df.pop("Nationality")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nat, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

foot = df.pop("Preferred Foot")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(foot, prefix='foot').reset_index(drop=True)], axis=1, sort=False)

bot = df.pop("Body Type")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(bot, prefix='bot').reset_index(drop=True)], axis=1, sort=False)

pos = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

wr = df.pop("Work Rate")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(wr, prefix='wr').reset_index(drop=True)], axis=1, sort=False)

df.head()

Age  Overall  Potential    Value   Wage  Special  International Reputation  \
0   31       94         94  €110.5M  €565K     2202                       5.0   
1   33       94         94     €77M  €405K     2228                       5.0   
2   26       92         93  €118.5M  €290K     2143                       5.0   
3   27       91         93     €72M  €260K     1471                       4.0   
4   27       91         92    €102M  €355K     2281                       4.0   

   Weak Foot  Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF  \
0        4.0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1        4.0          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2        5.0          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        3.0          1.0    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4        5.0          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0       13.0   
4  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   
4           61.0         87.0    94.0       79.0       88.0     68.0   

   StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0            28.0           26.0       6.0        11.0       15.0   
1            31.0           23.0       7.0        11.0       15.0   
2            24.0           33.0       9.0         9.0       15.0   
3            21.0           13.0      90.

Finalmente vamos a normalizar las columnas que a pesar de que son relevantes para el algoritmo, no están expresadas de forma óptima. 
Empezaremos por las cifras monetarias en las cuales hemos de multiplicar según si son millones (M) o miles (K). Además hemos de eliminar el símbolo del euro para que el algoritmo entienda bien el número.

In [218]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    if(x != x): return 0
    
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

In [219]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df["Release Clause"] = df["Release Clause"].apply(value_to_float)
df.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves Height  Weight    LS  \
0                       5.0        4.0          4.0    5'7  159lbs  88+2   
1                       5.0        4.0          5.0    6'2  183lbs  91+3   
2                       5.0        5.0          5.0    5'9  150lbs  84+3   
3                       4.0        3.0          1.0    6'4  168lbs   NaN   
4                       4.0        5.0          4.0   5'11  154lbs  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      15.0   

Siguiendo con el proceso anterior normalizamos también los pesos y las alturas, ya que una de ellas contiene la unidad (lbs) y la otra el carácter " ' " el cual debemos pasar a decimal normal.

In [220]:
def peso_a_float(x):
    """ 
    Convertimos el peso a un float
    
    """ 
    x = float(x.replace("lbs", ""))
    return x

In [221]:
def altura_a_float(x):
    """
    Convertimos la altura en un float
    
    """   
    x = x.split("'")
    
    if(len(x[1]) == 1): x = float(x[0]) + float(x[1])/10
    else:
        y = float(x[1])/100
        x = float(x[0])+y
        
    return x

In [222]:
df["Weight"] = df["Weight"].apply(peso_a_float)
df["Height"] = df["Height"].apply(altura_a_float)
df.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight    LS  \
0                       5.0        4.0          4.0    5.70   159.0  88+2   
1                       5.0        4.0          5.0    6.20   183.0  91+3   
2                       5.0        5.0          5.0    5.90   150.0  84+3   
3                       4.0        3.0          1.0    6.40   168.0   NaN   
4                       4.0        5.0          4.0    5.11   154.0  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      1

Una vez más vamos a arreglar las últimas columnas mal expresadas. En este caso las columnas de estadísticas para los jugadores las cuales tienen la forma de x+y. Esta suma representa las estadísticas del jugador más un modificador. He supuesto que el modificador depende de la predilección del jugador a ciertas estadísticas, así que sumar el valor no tendrá ningún impacto en el algoritmo y así conseguimos que pueda leerlo.

En este apartado hemos podido observar en clase que tenemos un problema con los porteros, ya que no tienen estadísticas registradas en muchos aspectos. La solución tomada ha sido hacer una media de las estadísticas y colocárselo a los porteros de forma que el algoritmo no los desheche automáticamente, ya que el no tener estadísticas de tiro (por ejemplo) no influye en que un portero pueda ser muy bueno y valer mucho más dinero que un buen tirador.

> - En primer lugar sumamos las estadísticas excluyendo las estadísticas NaN, las cuales dejaremos tal cual.
> - En segundo lugar sustituiremos los valores NaN por una media de las estadísticas mediante la función "replace".

In [223]:
def add_stats(x):
    """
    Suma las estadísticas y sustituye las columnas NaN por la media
    
    """

    if(x != x): return np.nan
        
    
    stats = x.split('+')
    
    
    return (float(stats[0]) + float(stats[1]))

In [224]:
stats = ["LS","ST","RS","LW","LF","CF","RF","RW","LAM","CAM","RAM","LM","LCM","CM","RCM","RM","LWB","LDM","CDM","RDM","RWB","LB","LCB","CB","RCB","RB"]
for st in stats: df[st] = df[st].apply(add_stats)
df.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight    LS  \
0                       5.0        4.0          4.0    5.70   159.0  90.0   
1                       5.0        4.0          5.0    6.20   183.0  94.0   
2                       5.0        5.0          5.0    5.90   150.0  87.0   
3                       4.0        3.0          1.0    6.40   168.0   NaN   
4                       4.0        5.0          4.0    5.11   154.0  85.0   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  90.0  90.0  94.0  95.0  95.0  95.0  94.0  95.0  95.0  95.0  93.0  86.0   
1  94.0  94.0  92.0  93.0  93.0  93.0  92.0  91.0  91.0  91.0  91.0  84.0   
2  87.0  87.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  91.0  84.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  85.0  85.0  90.0  90.0  90.0  90.0  90.0  91.0  91.0  91.0  91.0  90.0   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  86.0  86.0  93.0  66.0  63.0  63.0  63.0  66.0  61.0  49.0  49.0  49.0   
1  84.0  84.0  91.0  68.0  64.0  64.0  64.0  68.0  64.0  56.0  56.0  56.0   
2  84.0  84.0  91.0  68.0  63.0  63.0  63.0  68.0  63.0  50.0  50.0  50.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  90.0  90.0  91.0  80.0  80.0  80.0  80.0  80.0  76.0  69.0  69.0  69.0   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  61.0      84.0       95.0             70.0          90.0     86.0   
1  64.0      84.0       94.0             89.0          81.0     87.0   
2  63.0      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  76.0      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      1

In [225]:
stats = ["LS","ST","RS","LW","LF","CF","RF","RW","LAM","CAM","RAM","LM","LCM","CM","RCM","RM","LWB","LDM","CDM","RDM","RWB","LB","LCB","CB","RCB","RB"]
for st in stats: df[st] = df[st].replace([np.nan], df[st].mean())
df.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight  \
0                       5.0        4.0          4.0    5.70   159.0   
1                       5.0        4.0          5.0    6.20   183.0   
2                       5.0        5.0          5.0    5.90   150.0   
3                       4.0        3.0          1.0    6.40   168.0   
4                       4.0        5.0          4.0    5.11   154.0   

          LS         ST         RS         LW         LF         CF  \
0  90.000000  90.000000  90.000000  94.000000  95.000000  95.000000   
1  94.000000  94.000000  94.000000  92.000000  93.000000  93.000000   
2  87.000000  87.000000  87.000000  92.000000  92.000000  92.000000   
3  59.842647  59.842647  59.842647  61.066809  60.750471  60.750471   
4  85.000000  85.000000  85.000000  90.000000  90.000000  90.000000   

          RF         RW        LAM        CAM        RAM         LM  \
0  95.000000  94.000000  95.000000  95.000000  95.000000  93.000000   
1  93.000000  92.000000  91.000000  91.000000  91.000000  91.000000   
2  92.000000  92.000000  92.000000  92.000000  92.000000  91.000000   
3  60.750471  61.066809  61.015195  61.015195  61.015195  61.760455   
4  90.000000  90.000000  91.000000  91.000000  91.000000  91.000000   

         LCM         CM        RCM         RM        LWB        LDM  \
0  86.000000  86.000000  86.000000  93.000000  66.000000  63.000000   
1  84.000000  84.000000  84.000000  91.000000  68.000000  64.000000   
2  84.000000  84.000000  84.000000  91.000000  68.000000  63.000000   
3  60.244569  60.244569  60.244569  61.760455  59.516891  58.831659   
4  90.000000  90.000000  90.000000  91.000000  80.000000  80.000000   

         CDM        RDM        RWB         LB        LCB         CB  \
0  63.000000  63.000000  66.000000  61.000000  49.000000  49.000000   
1  64.000000  64.000000  68.000000  64.000000  56.000000  56.000000   
2  63.000000  63.000000  68.000000  63.000000  50.000000  50.000000   
3  58.831659  58.831659  59.516891  58.806229  57.614027  57.614027   
4  80.000000  80.000000  80.000000  76.000000  69.000000  69.000000   

         RCB         RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  49.000000  61.000000      84.0       95.0             70.0          90.0   
1  56.000000  64.000000      84.0       94.0             89.0          81.0   
2  50.000000  63.000000      79.0       87.0             62.0          84.0   
3  57.614027  58.806229      17.0       13.0             21.0          50.0   
4  69.000000  76.000000      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0      

## Predicción
La predicción del valor del jugador se hará entrenando un modelo de regresión lineal tal y como se vio en clase:

In [226]:
val = df.pop("Value")

Se utiliza 1/3 de la muestra para el test

In [227]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=31)

In [228]:
len(X_train)

12005

Entrenamos el modelo:

In [229]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

Finalmente obtenemos una métrica 𝑅2 para la regresión

In [230]:
preds = reg.predict(X_test)
r2_score(preds, y_test)

0.9643220475201197

Segun la métrica obtenemos un 96.4322% de fiabilidad lo cual es bastante aceptable.

In [231]:
y_test

4783     2400000.0
12232     290000.0
5615      600000.0
1176     8500000.0
11209     375000.0
           ...    
15517     190000.0
13134     525000.0
2986     3700000.0
15190     240000.0
6946     1200000.0
Name: Value, Length: 5913, dtype: float64

In [232]:
preds[0]

3009778.755317174

Finalmente si observamos la realidad con el dato predicho vemos que la diferencia no es demasiado grande teniendo en cuenta los valores utilizados.